### ISSUE TITLE
`cabal new-build` exits with exit-code 0 in case of compile failures

### ISSUE THREAD

**Issue by hvr** [haskell/cabal#3495]
iirc @dcoutts mentioned this was a known issue he was working, but I didn't find the issue; so I hope this one isn't a duplicate

This makes it difficult detect failures in `new-build` based CI jobs to, as reported in https://github.com/hvr/multi-ghc-travis/issues/52

**Comment by hvr** [comment]
Here's a totally unpolished hack:

```diff
diff --git a/cabal-install/Distribution/Client/CmdBuild.hs b/cabal-install/Distribution/Client/CmdBuild.hs
index 70e83ac..1f015e7 100644
--- a/cabal-install/Distribution/Client/CmdBuild.hs
+++ b/cabal-install/Distribution/Client/CmdBuild.hs
@@ -23,8 +23,12 @@ import Distribution.Simple.Setup
 import Distribution.Verbosity
          ( normal )
 
-import Control.Monad (unless)
+import Control.Monad (unless,forM_)
+import Distribution.Simple.Utils (die)
 
+import qualified Distribution.Client.InstallPlan as IP
+import Distribution.Text ( display )
+import Distribution.Package
 
 -- | The @build@ command does a lot. It brings the install plan up to date,
 -- selects that part of the plan needed by the given or implicit targets and
@@ -56,6 +60,14 @@ buildAction (configFlags, configExFlags, installFlags, haddockFlags)
       _plan <- runProjectBuildPhase
                  verbosity
                  buildCtx
+      -- print _plan
+      let failedPkgs = [ (srcpkg,failure) | IP.Failed srcpkg failure <- IP.toList _plan ]
+      forM_ failedPkgs $ \(srcpkg,failure) -> do
+          putStrLn $ display (packageId srcpkg) ++
+                     " (" ++ display (installedUnitId srcpkg) ++ "): " ++
+                     show failure
+
+      unless (null failedPkgs) $ die "build failed"
       --TODO: [required eventually] report on build failures in residual plan
       return ()
   where
```

Which results in the following verbose output:

```
In order, the following will be built (use -v for more details):
haskell-src-exts-1.13.5
hsx-0.10.4
Configuring haskell-src-exts-1.13.5...
Building haskell-src-exts-1.13.5...
haskell-src-exts-1.13.5 (haskell-src-exts-1.13.5-3174c81658ed1b1f3b09e95603d4a218507b2d04472d8f0f52b6b165f48b8002): BuildFailed "ExitFailure 1"
hsx-0.10.4 (hsx-0.10.4-inplace): DependentFailed (PackageIdentifier {pkgName = PackageName {unPackageName = "haskell-src-exts"}, pkgVersion = Version {versionBranch = [1,13,5], versionTags = []}})
cabal: build failed
```

@dcoutts what kind of output would you prefer?

**Comment by 23Skidoo** [comment]
Fixed by #3504.
